### PR TITLE
fix: skip invalid overlay coordinates

### LINKED ISSUE #128: Skip overlay descriptors with invalid coordinates in JS with a __DEV__ warning
## Problem

An overlay whose coordinates are `NaN` or out of range is forwarded to the native map
verbatim, and on Android the Google Maps SDK throws inside the Fabric mount transaction —
one bad row from an API response takes down a screen full of good overlays. #125 covers the
native guards; this issue is the JS half, which can land on its own.

The GeoJSON path already does exactly what is needed here. `parseGeojson.ts` requires finite
numbers (L16–21) and minimum ring lengths (L32–40), and skips bad features with a `__DEV__`
warning via `warnGeojson.ts`. Overlays passed as `<Marker>`, `<Polyline>`, `<Polygon>` and
`<Circle>` children get none of that — `useCollectedOverlays.ts` copies `coordinates`,
`center` and `radius` straight into the descriptor.

## Scope

TypeScript only. Deliberately **not** in scope:

- `region` / `initialRegion` — handled in #125, because making an invalid region `undefined`
  interacts with #119 (value props that become `undefined` on a mounted view).
- Native guards in Kotlin and Swift — #125.

## Suggested fix

**1. Add `package/src/overlays/validateOverlay.ts`**

```ts
export function isValidCoordinate(value: Coordinate | undefined): boolean
export function isValidCoordinateList(value: Coordinate[] | undefined, minimumLength: number): boolean
export function isValidRadius(value: number | undefined): boolean
```

Rules:

- a coordinate is valid when both fields pass `Number.isFinite`, `latitude` is within
  `[-90, 90]` and `longitude` within `[-180, 180]`
- a radius is valid when it is finite and `>= 0`
- minimum lengths: **2** for a polyline, **3** for a polygon

> ⚠️ Do not "harmonize" the polygon minimum with GeoJSON's. `parseGeojson.ts:54` uses **4**
> because a GeoJSON linear ring is closed (the last position repeats the first). `<Polygon>`
> coordinates are not closed — both providers close the ring themselves — so **3** is correct
> here. Two different numbers, same intent.

Out-of-range values are **rejected, not clamped**. Clamping is what Android's `LatLng` does
today and it silently renders the overlay in the wrong place, which is the behaviour #125
is trying to get rid of.

**2. Add `package/src/overlays/warnOverlay.ts`**

Copy the shape of `warnGeojson.ts` exactly — same `__DEV__` guard, same
`[react-native-better-maps]` prefix, different label:

```ts
export function warnOverlay(message: string): void {
  if ((globalThis as { __DEV__?: boolean }).__DEV__ !== true) {
    return;
  }
  console.warn(`[react-native-better-maps] Overlay: ${message}`);
}
```

**3. Wire it into the collectors in `package/src/hooks/useCollectedOverlays.ts`**

In the `Marker` (L74), `Polyline` (L109), `Polygon` (L132) and `Circle` (L156) collectors,
validate before the `state.*.push(...)` call. When a descriptor is invalid, warn once with
its resolved id and return without pushing — do not push a partial descriptor, and do not
touch `state.registry` or the `has*Press` flags for it.

Include the id and what failed in the message, e.g.
`` `polygon "zone-3" skipped: needs at least 3 valid coordinates` ``.

A bad overlay must never remove the good ones around it — that is the whole point.

## Tests

Add `package/src/overlays/__tests__/validateOverlay.ts` tests next to the existing ones,
covering: `NaN`, `Infinity`, latitude `±90.0001`, longitude `±180.0001`, the boundary values
themselves (`±90` / `±180` are valid), an empty array, an array one short of the minimum, and
a negative radius. `0` is a valid radius.

Then extend `package/src/overlays/__tests__/` with a collector test asserting that a
`<Polygon coordinates={[]} />` next to a valid polygon leaves the valid one in the collected
array.

`package/src/geojson/__tests__/warnGeojson.test.ts` shows how the `__DEV__` warning is
tested in this repo.

## Notes

Related: #125 (native guards and region validation).

### PR DESCRIPTION
**Pull request by TayfurYldz**
## Summary
- validate Marker/Polyline/Polygon/Circle coordinates before collecting descriptors
- reject non-finite/out-of-range coordinates and invalid circle radii without mutating callback state
- emit `__DEV__` warnings for skipped overlays so one bad overlay does not affect valid siblings
- add regression coverage for coordinate/radius boundaries, warning behavior, and invalid+valid polygon collection

Closes #128

## Validation
- `bun run --filter react-native-better-maps test` — 170 pass, 0 fail
- `bun run nitrogen`
- `bun run typecheck`
- `bun run lint`
- `bun run build`
- `git diff --check`

---
<a href="https://app.blacksmith.sh/gmi-software/codesmith/react-native-better-maps/pr/143"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with [code]smith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1792265196&installation_model_id=428715&pr_number=143&repository=gmi-software%2Freact-native-better-maps&return_to=https%3A%2F%2Fgithub.com%2Fgmi-software%2Freact-native-better-maps%2Fpull%2F143&signature=b44c7f2bc30bbd8734094cb6fe4cb0966cf8acd87102f6b83b6fb457446944b2"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with [code]smith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith-bot</code> with what you need. Autofix is disabled.</sup>

### CODE REVIEW

File: package/src/overlays/validateOverlay.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+export function isValidCoordinateList(
+  value: Coordinate[] | undefined,
+  minimumLength: number,
+): boolean {
+  return (
+    value != null &&
+    value.length >= minimumLength &&
+    value.every(isValidCoordinate)
```
[the remark this answers is not in the document]
Comment (TayfurYldz) [reply package/src/overlays/validateOverlay.ts]: Confirmed. `Array.prototype.every()` skipped sparse holes, so `new Array(2)` passed the previous validator. Updated the validator to require an actual array and validate each indexed entry explicitly, and added a regression test. Targeted tests pass 8/8; full package suite passes 171/171, with typecheck, lint, build, and `git diff --check` also clean.

**Review by jkasprzyk17 — CHANGES_REQUESTED** [review]
Thanks @TayfurYldz, this is a solid JS-side guard for #128. Finite/range checks, reject-don't-clamp, polygon min 3 (not GeoJSON's 4), the sparse-array loop, and skipping without touching registry / `has*Press` are all right. A bad overlay no longer takes the good ones with it.

Two changes before merge:

1. `collectOverlayChild` shouldn't be exported from `useCollectedOverlays.ts`. Overlay collectors already live in `package/src/overlays/` (`collectGeojsonOverlays` is the pattern). Move the collector table + dispatcher there, keep a one-line call in the hook, and import the test from `overlays/`. Then we can drop the `mock.module` / dynamic import. That's only there because the test is going through a hook module.

2. Document skip-on-invalid on the public overlay props (`MarkerProps.coordinate`, `PolylineProps.coordinates`, `PolygonProps.coordinates`, `CircleProps.center` / `radius`). `geojsonToOverlayDescriptors` already says invalid geometry is skipped with a development warning; overlay children should say the same. Callers need to know a NaN coordinate or a 2-point polygon is omitted, not clamped and not thrown.

Small nit: polyline/polygon always warn `needs at least N valid coordinates`, even when the list is long enough but a point is NaN / out of range. A generic "invalid coordinates" (or the actual reason) would be clearer.

`warnOverlay` as a copy of `warnGeojson` is fine, leave that. No need to throw instead of skip, and no native work in this PR.

Once those two are in, I'll approve ✅

**Comment by TayfurYldz** [comment]
Addressed both requested changes in `7fef058`:

- moved the overlay collector table/dispatcher into `src/overlays/collectOverlayChild.ts`; the hook now delegates to it, and the regression imports the overlay module directly without `mock.module` / dynamic import
- documented skip-on-invalid behavior on the public Marker/Polyline/Polygon/Circle props
- also changed the polyline/polygon warning to the generic `invalid coordinates` wording for malformed points

Validation: package tests 171/171 passed, typecheck passed, ESLint passed, build passed, and `git diff --check` is clean.

**Comment by jkasprzyk17** [comment]
Thanks again for the earlier fixes those look good.

#118 just landed on `main`, so this branch is now conflicting. Please rebase onto latest `main` and resolve the collector overlap.

When resolving, keep both sides:
- validation / skip-on-invalid from this PR
- `markerColor` + `zIndex` forwarding from #118 (`collectMarkerOverlay` on main)

A clean shape would be: validate the coordinate first, then call `collectMarkerOverlay` (or otherwise still forward `markerColor` / `zIndex` into the descriptor). After the rebase I’ll take another look.

**Comment by TayfurYldz** [comment]
Rebased onto current `main` after #118 and resolved the collector overlap in `491098f`. The marker path now validates first and delegates valid markers to main's `collectMarkerOverlay`, preserving `markerColor` and `zIndex` forwarding while invalid markers are still skipped before registry/press state is touched.

Post-rebase validation: focused overlap tests 7/7 passed; full package suite 181/181 passed; typecheck, ESLint, build, and `git diff --check` passed.

**Comment by jkasprzyk17** [comment]
Nice, rebase looks clean. Validate → `collectMarkerOverlay` keeps both sides of the overlap.

LGTM ✅

**Review by jkasprzyk17 — APPROVED** [review]
